{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7c3d839c",
   "metadata": {},
   "source": [
    "# <font color=green>*BCBIO 490*</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "647e2899",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import os\n",
    "import math\n",
    "from Bio import SeqIO\n",
    "import blosum as bl\n",
    "\n",
    "# The class that performs sequence alignments on two given sequences based on user-defined scores and penalties.\n",
    "class Sequence_Alignment:\n",
    "    \n",
    "    # Constructor method \n",
    "    # @param match : score used if two characters are identical\n",
    "    # @param mismatch : score used if two characters are not identical\n",
    "    # @param gapOpen : penalty for creating a gap '-' in an alignment\n",
    "    # @param gapExtend : penalty for creating a consecutive gap \n",
    "    # @param diff_block : penalty for a block consisting of alignments in which large regions of both sequences  \n",
    "    #                     are different. \n",
    "    # @param ambiguous : score used if one of the two characters is 'N', which is a sequencing error in which \n",
    "    #                    the particular nucleotide is unknown. \n",
    "    def __init__(self, match=10, mismatch=-20, gapOpen=40, gapExtend=2, diff_block=300, ambiguous=0):\n",
    "        self.gapOpen = abs(gapOpen)\n",
    "        self.gapExtend = abs(gapExtend)\n",
    "        self.diff = abs(diff_block)\n",
    "        self.sub_mat = np.zeros((129,129))\n",
    "        for i in range(len(self.sub_mat)):\n",
    "            for j in range(len(self.sub_mat[i])):\n",
    "                if i == 'N' or j == 'N':\n",
    "                    self.sub_mat[i][j] = 0 \n",
    "                elif i != j:\n",
    "                    self.sub_mat[i][j] = mismatch \n",
    "                else:\n",
    "                    self.sub_mat[i][j] = match\n",
    "                    \n",
    "        self.protein = bl.BLOSUM(62) # Scoring system for protein alignments\n",
    "    \n",
    "    # Used for accessing the scoring matrix generated in the constructor method given two characters\n",
    "    # @param char1 : first character\n",
    "    # @param char2 : second character\n",
    "    # @param mode : determines the scoring system that the algorithm will use. Default is 0, which is the \n",
    "    #               scoring system for DNA/ RNA sequences. If not 0, the algorithm uses the scoring system \n",
    "    #               for protein sequences. \n",
    "    def getScore(self, char1, char2, mode=0):\n",
    "        return self.sub_mat[ord(char1.upper())][ord(char2.upper())] if mode == 0 \\\n",
    "                else self.protein[char1.upper()][char2.upper()]\n",
    "    \n",
    "    # Modifies the gapOpen variable \n",
    "    def setGapOpenPenalty(self, penalty):\n",
    "        self.gapOpen = abs(penalty)\n",
    "    \n",
    "    # Modifies the gapExtend variable\n",
    "    def setGapExtensionPenalty(self, penalty):\n",
    "        self.gapExtend = abs(penalty)\n",
    "    \n",
    "    # Modifies the diff variable\n",
    "    def setDifferenceBlockPenalty(self, penalty):\n",
    "        self.diff = abs(penalty)\n",
    "    \n",
    "    # Method that takes in two FASTA files containing same type of sequences, and aligns them based on the \n",
    "    # alignment type specified by the user. \n",
    "    # @param fasta1 : file name of the first FASTA file \n",
    "    # @param fasta2 : file name of the second FASTA file\n",
    "    # @align_type : determines the alignment type. Possible values are 'global', 'local', or 'gap3'\n",
    "    def alignInput(self, fasta1, fasta2, align_type):\n",
    "        \"\"\"\n",
    "        cwd = os.getcwd()  # Get the current working directory (cwd)\n",
    "        files = os.listdir(cwd)  # Get all the files in that directory\n",
    "        print(\"Files in %r: %s\" % (cwd, files))\n",
    "        \"\"\"\n",
    "        if os.path.getsize(fasta1) == 0 or os.path.getsize(fasta2) == 0:\n",
    "            return \"Input contains empty file(s)\"\n",
    "        if align_type != 'global' and align_type != 'local' and align_type != 'gap3':\n",
    "            return \"This program only has three alignment types: 'global' or 'local' or 'gap3'\"\n",
    "                    \n",
    "        with open(fasta1, \"r\") as handle1, open(fasta2, \"r\") as handle2:\n",
    "            for record1, record2 in zip(SeqIO.parse(handle1, \"fasta\"), SeqIO.parse(handle2, \"fasta\")):\n",
    "                print(\"Aligning \" + record1.id + \" with \" + record2.id + \" using \" + align_type + \" alignment: \")\n",
    "                print()\n",
    "                if align_type == 'global':\n",
    "                    self.global_alignment(record1.seq, record2.seq)\n",
    "                elif align_type == 'local':\n",
    "                    self.local_alignment(record1.seq, record2.seq)\n",
    "                elif align_type == 'gap3':\n",
    "                    self.gap3(record1.seq, record2.seq)\n",
    "\n",
    "    # Method that allows user to manually create two sequences and align them based on the alignment type\n",
    "    # specified by the user.\n",
    "    # @param seq_type : determines the type of sequences that can be generated. Possible values are \n",
    "    #                   'DNA', 'RNA', or 'PROTEIN'\n",
    "    def manualAlign(self, seq_type):\n",
    "        seq = {'DNA': ('ATCG', 0), 'RNA': ('AUCG', 0), 'PROTEIN': ('AGVLISTCMPFYWDENQKRH', 1)}\n",
    "        if seq_type.upper() not in seq:\n",
    "            print(\"Choose either 'DNA', 'RNA', or 'PROTEIN' sequences. No other types of sequences are allowed.\")\n",
    "            return \n",
    "\n",
    "        while True:\n",
    "            seq1 = input(\"Enter a \" + seq_type.upper() + \" sequence to be aligned: \")\n",
    "            validation = [i.upper() in seq[seq_type.upper()][0] for i in seq1]\n",
    "            if (all(validation)):\n",
    "                break\n",
    "            print(\"***\" + seq_type.upper() + \"sequences should contain only these characters: \" + \\\n",
    "                  seq[seq_type.upper()][0] + \", and no spaces are allowed ! ***\")\n",
    "        \n",
    "        print(\"------------------------------------------------------------------------------------------------\")\n",
    "        \n",
    "        while True:\n",
    "            seq2 = input(\"Enter another \" + seq_type.upper() + \" sequence to be aligned: \")\n",
    "            validation = [i.upper() in seq[seq_type.upper()][0] for i in seq2]\n",
    "            if (all(validation)):\n",
    "                break\n",
    "            print(\"***\" + seq_type.upper() + \"sequences should contain only these characters: \" + \\\n",
    "                  seq[seq_type.upper()][0] + \", and no spaces are allowed ! ***\")\n",
    "            print()\n",
    "        \n",
    "        print(\"------------------------------------------------------------------------------------------------\")\n",
    "        \n",
    "        print(\"Please choose an alignment mode: \")\n",
    "        print(\"\\t\" + \"1. Type 'global' for global alignment\")\n",
    "        print(\"\\t\" + \"2. Type 'local' for local alignment\")\n",
    "        print(\"\\t\" + \"3. Type 'gap3' for gap3 alignment\")\n",
    "        print()\n",
    "        \n",
    "        while True:\n",
    "            align = input(\"Enter your desired alignment type: \")\n",
    "            if align.lower() == 'global' or align.lower() == 'local' or align.lower() == 'gap3':\n",
    "                break\n",
    "            print(\"***Please select alignment type only from 'global' or 'local or 'gap3' ! ***\")\n",
    "            print()\n",
    "            \n",
    "        print(\"------------------------------------------------------------------------------------------------\")\n",
    "        print()\n",
    "        print(\"Outputting \" + align.lower() + \" Alignment ...\")\n",
    "        print()\n",
    "        if align.lower() == 'local':\n",
    "            self.local_alignment(seq1, seq2, seq[seq_type.upper()][1])\n",
    "        elif align.lower() == 'global':\n",
    "            self.global_alignment(seq1, seq2, seq[seq_type.upper()][1])\n",
    "        else:\n",
    "            self.gap3(seq1, seq2, seq[seq_type.upper()][1])\n",
    "    \n",
    "    # Traceback method for generating the global alignment of two given sequences, which is only called by\n",
    "    # the method global_alignment_trace when the first sequence has a size below a certain threshold. \n",
    "    # @param seq1 : first sequence\n",
    "    # @param seq2 : second sequence\n",
    "    # @param S : the score used as the basis of the construction of s_mat\n",
    "    # @param D : the score used as the basis of the construction of d_mat\n",
    "    # @param curr : the matrix to start from when generating the optimal alignment\n",
    "    # @param mode : determines the scoring system that the algorithm will use. 0 is the \n",
    "    #               scoring system for DNA/ RNA sequences. If not 0, the algorithm uses the scoring system \n",
    "    #               for protein sequences. \n",
    "    def global_traceback(self, seq1, seq2, S, D, curr, mode):\n",
    "        M = len(seq1)\n",
    "        N = len(seq2)\n",
    "        \n",
    "        seq1 = \" \" + seq1\n",
    "        seq2 = \" \" + seq2\n",
    "        \n",
    "        s_mat = np.zeros((M + 1, N + 1))\n",
    "        i_mat = np.zeros((M + 1, N + 1))      # Generation of score matrices using suffixes of both sequences\n",
    "        d_mat = np.zeros((M + 1, N + 1))\n",
    "        \n",
    "        s_mat[M][N] = S\n",
    "        d_mat[M][N] = D\n",
    "        i_mat[M][N] = s_mat[M][N] - self.gapOpen\n",
    "        \n",
    "        for i in range(N - 1, -1, -1):\n",
    "            i_mat[M][i] = i_mat[M][i + 1] - self.gapExtend\n",
    "            s_mat[M][i] = i_mat[M][i]\n",
    "            d_mat[M][i] = s_mat[M][i] - self.gapOpen\n",
    "        \n",
    "        for j in range(M - 1, -1, -1):\n",
    "            d_mat[j][N] = d_mat[j + 1][N] - self.gapExtend\n",
    "            s_mat[j][N] = d_mat[j][N]\n",
    "            i_mat[j][N] = s_mat[j][N] - self.gapOpen\n",
    "            \n",
    "            for k in range(N - 1, -1, -1):\n",
    "                d_mat[j][k] = max(d_mat[j + 1][k] - self.gapExtend, s_mat[j + 1][k] - self.gapOpen - self.gapExtend)\n",
    "                i_mat[j][k] = max(i_mat[j][k + 1] - self.gapExtend, s_mat[j][k + 1] - self.gapOpen - self.gapExtend)\n",
    "                s_mat[j][k] = max(s_mat[j + 1][k + 1] + self.getScore(seq1[j + 1], seq2[k + 1], mode), d_mat[j][k], i_mat[j][k])\n",
    "                \n",
    "        current = curr\n",
    "        i = j = 0\n",
    "        align_seq1 = \"\"\n",
    "        align_seq2 = \"\"\n",
    "        align_mid = \"\"\n",
    "\n",
    "        while i <= M and j <= N:\n",
    "            if current == 'S':\n",
    "                if i == M and j == N:\n",
    "                    break\n",
    "                elif i == M or s_mat[i][j] == i_mat[i][j]:\n",
    "                    current = 'I'\n",
    "                    continue\n",
    "                elif j == N or s_mat[i][j] == d_mat[i][j]:\n",
    "                    current = 'D'\n",
    "                    continue\n",
    "                    \n",
    "                align_seq1 += seq1[i + 1]\n",
    "                align_seq2 += seq2[j + 1]\n",
    "                if seq1[i + 1].upper() != seq2[j + 1].upper():\n",
    "                    if self.getScore(seq1[i + 1].upper(), seq2[j + 1].upper(), mode) > 0:\n",
    "                        align_mid += \"=\"\n",
    "                    else:\n",
    "                        align_mid += \"*\"\n",
    "                else:\n",
    "                    align_mid += \"|\"\n",
    "                i += 1\n",
    "                j += 1\n",
    "                continue\n",
    "                \n",
    "            elif current == 'I':\n",
    "                align_seq1 += '-'\n",
    "                align_seq2 += seq2[j + 1]\n",
    "                align_mid += \" \"\n",
    "                if (j == N - 1) or (i_mat[i][j] == s_mat[i][j + 1] - self.gapOpen - self.gapExtend):\n",
    "                    current = 'S'\n",
    "                j += 1\n",
    "                continue\n",
    "                \n",
    "            elif current == 'D':\n",
    "                align_seq1 += seq1[i + 1]\n",
    "                align_seq2 += '-'\n",
    "                align_mid += \" \"\n",
    "                if (i == M - 1) or (d_mat[i][j] == s_mat[i + 1][j] - self.gapOpen - self.gapExtend):\n",
    "                    current = 'S'\n",
    "                i += 1\n",
    "                continue\n",
    "        \n",
    "        return [s_mat[0][0], align_seq1, align_mid, align_seq2]\n",
    "    \n",
    "    # Divide-and-Conquer method for generating the global alignment of two given sequences, which allows for\n",
    "    # the analysis of large datasets. \n",
    "    # @param seq1 : first sequence\n",
    "    # @param seq2 : second sequence\n",
    "    # @param s : the score used as the basis of the construction of s_mat\n",
    "    # @param d : the score used as the basis of the construction of d_mat\n",
    "    # @param S : the score used as the basis of the construction of S_MAT\n",
    "    # @param D : the score used as the basis of the construction of D_MAT\n",
    "    # @param curr : the matrix to start from when generating the optimal alignment\n",
    "    # @param mode : determines the scoring system that the algorithm will use. 0 is the \n",
    "    #               scoring system for DNA/ RNA sequences. If not 0, the algorithm uses the scoring system \n",
    "    #               for protein sequences. \n",
    "    def global_alignment_trace(self, seq1, seq2, s, d, S, D, curr, mode):\n",
    "        \n",
    "        # When the first sequence is less than 50 characters, call global_traceback method\n",
    "        if len(seq1) <= 50: \n",
    "            arr = self.global_traceback(seq1, seq2, S, D, curr, mode)\n",
    "            return arr\n",
    "        \n",
    "        else:\n",
    "            M = len(seq1)\n",
    "            N = len(seq2)\n",
    "            \n",
    "            # Divide the first sequence into half\n",
    "            imid = M // 2\n",
    "        \n",
    "            seq1 = \" \" + seq1\n",
    "            seq2 = \" \" + seq2\n",
    "            s_row = np.zeros(N + 1)\n",
    "            i_row = np.zeros(N + 1)           # Generation of score array using prefixes of both sequences\n",
    "            d_row = np.zeros(N + 1)\n",
    "        \n",
    "            S_ROW = np.zeros(N + 1)\n",
    "            I_ROW = np.zeros(N + 1)           # Generation of score array using suffixes of both sequences\n",
    "            D_ROW = np.zeros(N + 1)\n",
    "        \n",
    "            s_row[0] = s\n",
    "            i_row[0] = s_row[0] - self.gapOpen\n",
    "            d_row[0] = d\n",
    "        \n",
    "            S_ROW[N] = S\n",
    "            I_ROW[N] = S_ROW[N] - self.gapOpen\n",
    "            D_ROW[N] = D\n",
    "            \n",
    "            for j in range(1, N + 1):\n",
    "                i_row[j] = i_row[j-1] - self.gapExtend\n",
    "                s_row[j] = i_row[j]\n",
    "                d_row[j] = s_row[j] - self.gapOpen\n",
    "            \n",
    "            for i in range(1, imid + 1):\n",
    "                s_prev = s_row[0]\n",
    "            \n",
    "                d_row[0] = d_row[0] - self.gapExtend\n",
    "                s_row[0] = d_row[0]\n",
    "                i_row[0] = s_row[0] - self.gapOpen\n",
    "            \n",
    "                for j in range(1, N + 1):\n",
    "                    d_row[j] = max(d_row[j] - self.gapExtend, s_row[j] - self.gapOpen - self.gapExtend)\n",
    "                    i_row[j] = max(i_row[j - 1] - self.gapExtend, s_row[j - 1] - self.gapOpen - self.gapExtend)\n",
    "                    tmp = s_row[j]\n",
    "                    s_row[j] = max(s_prev + self.getScore(seq1[i], seq2[j], mode), d_row[j], i_row[j])\n",
    "                    s_prev = tmp\n",
    "\n",
    "            for j in range(N - 1, -1, -1):\n",
    "                I_ROW[j] = I_ROW[j + 1] - self.gapExtend\n",
    "                S_ROW[j] = I_ROW[j]\n",
    "                D_ROW[j] = S_ROW[j] - self.gapOpen\n",
    "            \n",
    "            for i in range(M - 1, imid - 1, -1):\n",
    "                S_PREV = S_ROW[N]\n",
    "            \n",
    "                D_ROW[N] = D_ROW[N] - self.gapExtend\n",
    "                S_ROW[N] = D_ROW[N]\n",
    "                I_ROW[N] = S_ROW[N] - self.gapOpen\n",
    "            \n",
    "                for j in range(N - 1, -1, -1):\n",
    "                    D_ROW[j] = max(D_ROW[j] - self.gapExtend, S_ROW[j] - self.gapOpen - self.gapExtend)\n",
    "                    I_ROW[j] = max(I_ROW[j + 1] - self.gapExtend, S_ROW[j + 1] - self.gapOpen - self.gapExtend)\n",
    "                    tmp = S_ROW[j]\n",
    "                    S_ROW[j] = max(S_PREV + self.getScore(seq1[i + 1], seq2[j + 1], mode), D_ROW[j], I_ROW[j])\n",
    "                    S_PREV = tmp\n",
    "                    \n",
    "            df = -math.inf\n",
    "            st = -math.inf\n",
    "        \n",
    "            jd = 0\n",
    "            js = 0\n",
    "            \n",
    "            # Find the point of intersection in which optimal alignment crosses to locate the partition point\n",
    "            # of the second sequence. \n",
    "            for j in range(N + 1):\n",
    "                if d_row[j] + D_ROW[j] + self.gapOpen > df:\n",
    "                    df = d_row[j] + D_ROW[j] + self.gapOpen\n",
    "                    jd = j\n",
    "                if s_row[j] + S_ROW[j] > st:\n",
    "                    st = s_row[j] + S_ROW[j]\n",
    "                    js = j\n",
    "\n",
    "            if max(df, st) == st:\n",
    "                jmid = js\n",
    "                final_score = st\n",
    "                current = 'S'\n",
    "                \n",
    "            else:\n",
    "                jmid = jd \n",
    "                final_score = df\n",
    "                current = 'D'\n",
    "            \n",
    "            # Recursively call itself on the partitions of both sequences. \n",
    "            top_align = self.global_alignment_trace(seq1[1: imid + 1], seq2[1:jmid + 1], s, d, \n",
    "                                         S_ROW[jmid], D_ROW[jmid], curr, mode)\n",
    "        \n",
    "            bottom_align = self.global_alignment_trace(seq1[imid+1::], seq2[jmid+1::], s_row[jmid], \n",
    "                                            d_row[jmid], S, D, current, mode)\n",
    "            \n",
    "            # Merge the two generated optimal alignments together to solve the optimal alignment of the \n",
    "            # whole sequences.\n",
    "            final_align_seq1 = top_align[1] + bottom_align[1]\n",
    "            final_align_mid = top_align[2] + bottom_align[2]\n",
    "            final_align_seq2 = top_align[3] + bottom_align[3]\n",
    "        \n",
    "            return [final_score, final_align_seq1, final_align_mid, final_align_seq2]\n",
    "        \n",
    "    # Method for global alignment, which aligns both given sequences end to end, and outputs the whole \n",
    "    # alignment. \n",
    "    # @param seq1 : first sequence\n",
    "    # @param seq2 : second sequence \n",
    "    # @param mode : determines the scoring system that the algorithm will use. Default is 0, which is the \n",
    "    #               scoring system for DNA/ RNA sequences. If not 0, the algorithm uses the scoring system \n",
    "    #               for protein sequences. \n",
    "    def global_alignment(self, seq1, seq2, mode=0):\n",
    "        s = 0\n",
    "        d = -self.gapOpen\n",
    "        if len(seq1) == 0 or len(seq2) == 0:\n",
    "            print(\"One of the sequences is empty.\")\n",
    "            return\n",
    "        \n",
    "        arr = self.global_alignment_trace(seq1, seq2, s, d, s, d, 'S', mode)\n",
    "        self.printAlignment(seq1, seq2, 0, 0, len(seq1), len(seq2), arr[0], arr[1], arr[2], arr[3])\n",
    "\n",
    "    # Method for local alignment, which aligns one sequence to the other at regions where there is high\n",
    "    # similarity, and outputs the region aligned. \n",
    "    # @param seq1 : first sequence\n",
    "    # @param seq2 : second sequence \n",
    "    # @param mode : determines the scoring system that the algorithm will use. Default is 0, which is the \n",
    "    #               scoring system for DNA/ RNA sequences. If not 0, the algorithm uses the scoring system \n",
    "    #               for protein sequences. \n",
    "    def local_alignment(self, seq1, seq2, mode=0):\n",
    "        if len(seq1) == 0 or len(seq2) == 0:\n",
    "            return\n",
    "        \n",
    "        highestScore = 0\n",
    "        firstRow = len(seq1)\n",
    "        firstCol = len(seq2)\n",
    "        \n",
    "        s_mat = np.zeros((len(seq1) + 1, len(seq2) + 1))\n",
    "        i_mat = np.zeros((len(seq1) + 1, len(seq2) + 1)) # Generation of score array using suffixes of both sequences\n",
    "        d_mat = np.zeros((len(seq1) + 1, len(seq2) + 1))\n",
    "        \n",
    "        d_mat[len(seq1)][len(seq2)] = -(self.gapOpen + self.gapExtend)\n",
    "        i_mat[len(seq1)][len(seq2)] = -(self.gapOpen + self.gapExtend)\n",
    "        for i in range(len(seq2) - 1, -1, -1):\n",
    "            i_mat[len(seq1)][i] =  -(self.gapOpen + self.gapExtend)\n",
    "            d_mat[len(seq1)][i] =  -(self.gapOpen + self.gapExtend)\n",
    "        \n",
    "        for j in range(len(seq1) - 1, -1, -1):\n",
    "            d_mat[j][len(seq2)] =  -(self.gapOpen + self.gapExtend)\n",
    "            i_mat[j][len(seq2)] =  -(self.gapOpen + self.gapExtend)\n",
    "        \n",
    "            for k in range(len(seq2) - 1, -1, -1):\n",
    "                d_mat[j][k] = max(d_mat[j + 1][k] - self.gapExtend, s_mat[j + 1][k] - self.gapOpen - self.gapExtend)\n",
    "                i_mat[j][k] = max(i_mat[j][k + 1] - self.gapExtend, s_mat[j][k + 1] - self.gapOpen - self.gapExtend)\n",
    "                s_mat[j][k] = max(0, s_mat[j + 1][k + 1] + self.getScore(seq1[j], seq2[k], mode), d_mat[j][k], i_mat[j][k])\n",
    "                \n",
    "                if highestScore < s_mat[j][k]:\n",
    "                    highestScore = s_mat[j][k]\n",
    "                    firstRow = j\n",
    "                    firstCol = k\n",
    "                    \n",
    "        current = 'S'\n",
    "        i = firstRow\n",
    "        j = firstCol \n",
    "        align_seq1 = \"\"\n",
    "        align_seq2 = \"\"\n",
    "        align_mid = \"\"\n",
    "\n",
    "        while i <= len(seq1) and j <= len(seq2):\n",
    "            if current == 'S':\n",
    "                if i == len(seq1) or j == len(seq2) or s_mat[i][j] == 0:\n",
    "                    break\n",
    "                elif s_mat[i][j] == i_mat[i][j]:\n",
    "                    current = 'I'\n",
    "                    continue\n",
    "                elif s_mat[i][j] == d_mat[i][j]:\n",
    "                    current = 'D'\n",
    "                    continue\n",
    "                align_seq1 += seq1[i]\n",
    "                align_seq2 += seq2[j]\n",
    "                if seq1[i].upper() != seq2[j].upper():\n",
    "                    if self.getScore(seq1[i].upper(), seq2[j].upper(), mode) > 0:\n",
    "                        align_mid += \"=\"\n",
    "                    else:\n",
    "                        align_mid += \"*\"\n",
    "                else:\n",
    "                    align_mid += \"|\"\n",
    "                i += 1\n",
    "                j += 1\n",
    "                continue\n",
    "                \n",
    "            elif current == 'I':\n",
    "                align_seq1 += '-'\n",
    "                align_seq2 += seq2[j]\n",
    "                align_mid += \" \"\n",
    "                if (j == len(seq2) - 1) or (i_mat[i][j] == s_mat[i][j + 1] - self.gapOpen - self.gapExtend):\n",
    "                    current = 'S'\n",
    "                j += 1\n",
    "                continue\n",
    "                \n",
    "            elif current == 'D':\n",
    "                align_seq1 += seq1[i]\n",
    "                align_seq2 += '-'\n",
    "                align_mid += \" \"\n",
    "                if (i == len(seq1) - 1) or (d_mat[i][j] == s_mat[i + 1][j] - self.gapOpen - self.gapExtend):\n",
    "                    current = 'S'\n",
    "                i += 1\n",
    "                continue \n",
    "                \n",
    "        lastRow = i\n",
    "        lastCol = j\n",
    "        score = s_mat[firstRow][firstCol]\n",
    "        \n",
    "        self.printAlignment(seq1, seq2, firstRow, firstCol, lastRow, lastCol, score, align_seq1, align_mid, align_seq2)\n",
    "    \n",
    "    # Traceback method for generating the gap3 alignment of two given sequences, which is only called by\n",
    "    # the method gap3_trace when the first sequence has a size below a certain threshold. \n",
    "    # @param seq1 : first sequence\n",
    "    # @param seq2 : second sequence\n",
    "    # @param S : the score used as the basis of the construction of s_mat\n",
    "    # @param D : the score used as the basis of the construction of d_mat\n",
    "    # @param H : the score used as the basis of the construction of h_mat\n",
    "    # @param curr : the matrix to start from when generating the optimal alignment\n",
    "    # @param mode : determines the scoring system that the algorithm will use. 0 is the \n",
    "    #               scoring system for DNA/ RNA sequences. If not 0, the algorithm uses the scoring system \n",
    "    #               for protein sequences. \n",
    "    def gap3_traceback(self, seq1, seq2, S, D, H, curr, mode):\n",
    "        M = len(seq1)\n",
    "        N = len(seq2)\n",
    "        \n",
    "        seq1 = \" \" + seq1\n",
    "        seq2 = \" \" + seq2\n",
    "        \n",
    "        s_mat = np.zeros((M + 1, N + 1))\n",
    "        i_mat = np.zeros((M + 1, N + 1))  # Generation of score array using suffixes of both sequences\n",
    "        d_mat = np.zeros((M + 1, N + 1))\n",
    "        h_mat = np.zeros((M + 1, N + 1))\n",
    "        \n",
    "        s_mat[M][N] = S \n",
    "        d_mat[M][N] = D \n",
    "        i_mat[M][N] = s_mat[M][N] - self.gapOpen\n",
    "        h_mat[M][N] = H\n",
    "        \n",
    "        for i in range(N - 1, -1, -1):\n",
    "            i_mat[M][i] = i_mat[M][i + 1] - self.gapExtend\n",
    "            h_mat[M][i] = H \n",
    "            s_mat[M][i] = max(i_mat[M][i], h_mat[M][i])\n",
    "            d_mat[M][i] = s_mat[M][i] - self.gapOpen\n",
    "        \n",
    "        for j in range(M - 1, -1, -1):\n",
    "            d_mat[j][N] = d_mat[j + 1][N] - self.gapExtend\n",
    "            h_mat[j][N] = H \n",
    "            s_mat[j][N] = max(d_mat[j][N], h_mat[j][N])\n",
    "            i_mat[j][N] = s_mat[j][N] - self.gapOpen\n",
    "            \n",
    "            for k in range(N - 1, -1, -1):\n",
    "                d_mat[j][k] = max(d_mat[j + 1][k] - self.gapExtend, s_mat[j + 1][k] - self.gapOpen - self.gapExtend)\n",
    "                i_mat[j][k] = max(i_mat[j][k + 1] - self.gapExtend, s_mat[j][k + 1] - self.gapOpen - self.gapExtend)\n",
    "                h_mat[j][k] = max(h_mat[j + 1][k], h_mat[j][k + 1], s_mat[j + 1][k] - self.diff, s_mat[j][k + 1] - self.diff)\n",
    "                s_mat[j][k] = max(s_mat[j + 1][k + 1] + self.getScore(seq1[j + 1], seq2[k + 1], mode), d_mat[j][k], i_mat[j][k], h_mat[j][k])\n",
    "                \n",
    "        current = curr\n",
    "        i = j = 0\n",
    "        align_seq1 = \"\"\n",
    "        align_seq2 = \"\"\n",
    "        align_mid = \"\"\n",
    "        \n",
    "        while i <= M and j <= N:\n",
    "            if current == 'S':\n",
    "                if i == M and j == N:\n",
    "                    break\n",
    "                elif s_mat[i][j] == h_mat[i][j]:\n",
    "                    current = 'H'\n",
    "                    continue\n",
    "                elif i == M or s_mat[i][j] == i_mat[i][j]:\n",
    "                    current = 'I'\n",
    "                    continue\n",
    "                elif j == N or s_mat[i][j] == d_mat[i][j]:\n",
    "                    current = 'D'\n",
    "                    continue\n",
    "                    \n",
    "                align_seq1 += seq1[i + 1]\n",
    "                align_seq2 += seq2[j + 1]\n",
    "                if seq1[i + 1].upper() != seq2[j + 1].upper():\n",
    "                    if self.getScore(seq1[i + 1].upper(), seq2[j + 1].upper(), mode) > 0:\n",
    "                        align_mid += \"=\"\n",
    "                    else:\n",
    "                        align_mid += \"*\"\n",
    "                else:\n",
    "                    align_mid += \"|\"\n",
    "                i += 1\n",
    "                j += 1\n",
    "                continue\n",
    "                \n",
    "            elif current == 'I':\n",
    "                align_seq1 += '-'\n",
    "                align_seq2 += seq2[j + 1]\n",
    "                align_mid += \" \"\n",
    "                if (j == N - 1) or (i_mat[i][j] == s_mat[i][j + 1] - self.gapOpen - self.gapExtend):\n",
    "                    current = 'S'\n",
    "                j += 1\n",
    "                continue\n",
    "                \n",
    "            elif current == 'D':\n",
    "                align_seq1 += seq1[i + 1]\n",
    "                align_seq2 += '-'\n",
    "                align_mid += \" \"\n",
    "                if (i == M - 1) or (d_mat[i][j] == s_mat[i + 1][j] - self.gapOpen - self.gapExtend):\n",
    "                    current = 'S'\n",
    "                i += 1\n",
    "                continue\n",
    "                \n",
    "            elif current == 'H':\n",
    "                if i == M and j == N:\n",
    "                    break\n",
    "                if i + 1 <= M:\n",
    "                    if h_mat[i][j] == h_mat[i + 1][j] or h_mat[i][j] == s_mat[i + 1][j] - self.diff:\n",
    "                        align_seq1 += seq1[i + 1]\n",
    "                        align_seq2 += \" \"\n",
    "                        align_mid += \"+\"\n",
    "                        if h_mat[i][j] == s_mat[i + 1][j] - self.diff:\n",
    "                            current = 'S'\n",
    "                        i += 1\n",
    "                        continue\n",
    "                        \n",
    "                if j + 1 <= N:\n",
    "                    if h_mat[i][j] == h_mat[i][j + 1] or h_mat[i][j] == s_mat[i][j + 1] - self.diff:\n",
    "                        align_seq1 += \" \"\n",
    "                        align_seq2 += seq2[j + 1]\n",
    "                        align_mid += \"+\"\n",
    "                        if h_mat[i][j] == s_mat[i][j + 1] - self.diff:\n",
    "                            current = 'S'\n",
    "                        j += 1\n",
    "                        continue\n",
    "\n",
    "        return [s_mat[0][0], align_seq1, align_mid, align_seq2]\n",
    "    \n",
    "    # Divide-and-Conquer method for generating the gap3 alignment of two given sequences, which allows for\n",
    "    # the analysis of large datasets. \n",
    "    # @param seq1 : first sequence\n",
    "    # @param seq2 : second sequence\n",
    "    # @param s : the score used as the basis of the construction of s_mat\n",
    "    # @param d : the score used as the basis of the construction of d_mat\n",
    "    # @param h : the score used as the basis of the construction of h_mat\n",
    "    # @param S : the score used as the basis of the construction of S_MAT\n",
    "    # @param D : the score used as the basis of the construction of D_MAT\n",
    "    # @param H : the score used as the basis of the construction of H_MAT\n",
    "    # @param curr : the matrix to start from when generating the optimal alignment\n",
    "    # @param mode : determines the scoring system that the algorithm will use. 0 is the \n",
    "    #               scoring system for DNA/ RNA sequences. If not 0, the algorithm uses the scoring system \n",
    "    #               for protein sequences. \n",
    "    def gap3_trace(self, seq1, seq2, s, d, h, S, D, H, curr, mode):\n",
    "        \n",
    "        # When the first sequence is less than 50 characters, call gap3_traceback method\n",
    "        if len(seq1) <= 50:\n",
    "            arr = self.gap3_traceback(seq1, seq2, S, D, H, curr, mode)\n",
    "            return arr\n",
    "        else: \n",
    "            M = len(seq1)\n",
    "            N = len(seq2)\n",
    "            imid = M // 2\n",
    "            \n",
    "            seq1 = \" \" + seq1\n",
    "            seq2 = \" \" + seq2\n",
    "            s_row = np.zeros(N + 1)\n",
    "            i_row = np.zeros(N + 1)   # Generation of score array using prefixes of both sequences\n",
    "            d_row = np.zeros(N + 1)   \n",
    "            h_row = np.zeros(N + 1)\n",
    "        \n",
    "            S_ROW = np.zeros(N + 1)\n",
    "            I_ROW = np.zeros(N + 1)   # Generation of score array using suffixes of both sequences\n",
    "            D_ROW = np.zeros(N + 1)\n",
    "            H_ROW = np.zeros(N + 1)\n",
    "        \n",
    "            s_row[0] = s\n",
    "            i_row[0] = s_row[0] - self.gapOpen\n",
    "            d_row[0] = d\n",
    "            h_row[0] = h\n",
    "        \n",
    "            S_ROW[N] = S\n",
    "            I_ROW[N] = S_ROW[N] - self.gapOpen\n",
    "            H_ROW[N] = H\n",
    "            D_ROW[N] = D\n",
    "        \n",
    "            for j in range(1, N + 1):\n",
    "                i_row[j] = i_row[j-1] - self.gapExtend\n",
    "                h_row[j] = h\n",
    "                s_row[j] = max(i_row[j], h_row[j])\n",
    "                d_row[j] = s_row[j] - self.gapOpen\n",
    "\n",
    "            for i in range(1, imid + 1):\n",
    "                s_prev = s_row[0]\n",
    "            \n",
    "                d_row[0] = d_row[0] - self.gapExtend\n",
    "                h_row[0] = h\n",
    "                s_row[0] = max(d_row[0], h_row[0])\n",
    "                i_row[0] = s_row[0] - self.gapOpen\n",
    "            \n",
    "                for j in range(1, N + 1):\n",
    "                    d_row[j] = max(d_row[j] - self.gapExtend, s_row[j] - self.gapOpen - self.gapExtend)\n",
    "                    i_row[j] = max(i_row[j - 1] - self.gapExtend, s_row[j - 1] - self.gapOpen - self.gapExtend)\n",
    "                    h_row[j] = max(h_row[j - 1], h_row[j], s_row[j - 1] - self.diff, s_row[j] - self.diff)\n",
    "                    tmp = s_row[j]\n",
    "                    s_row[j] = max(s_prev + self.getScore(seq1[i], seq2[j], mode), d_row[j], i_row[j], h_row[j])\n",
    "                    s_prev = tmp\n",
    "\n",
    "            for j in range(N - 1, -1, -1):\n",
    "                I_ROW[j] = I_ROW[j + 1] - self.gapExtend\n",
    "                H_ROW[j] = H\n",
    "                S_ROW[j] = max(I_ROW[j], H_ROW[j])\n",
    "                D_ROW[j] = S_ROW[j] - self.gapOpen\n",
    "\n",
    "            for i in range(M - 1, imid - 1, -1):\n",
    "                S_PREV = S_ROW[N]\n",
    "            \n",
    "                D_ROW[N] = D_ROW[N] - self.gapExtend\n",
    "                H_ROW[N] = H\n",
    "                S_ROW[N] = max(D_ROW[N], H_ROW[N])\n",
    "                I_ROW[N] = S_ROW[N] - self.gapOpen\n",
    "            \n",
    "                for j in range(N - 1, -1, -1):\n",
    "                    D_ROW[j] = max(D_ROW[j] - self.gapExtend, S_ROW[j] - self.gapOpen - self.gapExtend)\n",
    "                    I_ROW[j] = max(I_ROW[j + 1] - self.gapExtend, S_ROW[j + 1] - self.gapOpen - self.gapExtend)\n",
    "                    H_ROW[j] = max(H_ROW[j + 1], H_ROW[j], S_ROW[j + 1] - self.diff, S_ROW[j] - self.diff)\n",
    "                    tmp = S_ROW[j]\n",
    "                    S_ROW[j] = max(S_PREV + self.getScore(seq1[i + 1], seq2[j + 1], mode), D_ROW[j], I_ROW[j], H_ROW[j])\n",
    "                    S_PREV = tmp\n",
    "        \n",
    "            hk = -math.inf\n",
    "            df = -math.inf\n",
    "            st = -math.inf\n",
    "        \n",
    "            jh = 0\n",
    "            jd = 0\n",
    "            js = 0\n",
    "        \n",
    "            # Find the point of intersection in which optimal alignment crosses to locate the partition point\n",
    "            # of the second sequence.\n",
    "            for j in range(N + 1):\n",
    "                if h_row[j] + H_ROW[j] + self.diff > hk:\n",
    "                    hk = h_row[j] + H_ROW[j] + self.diff\n",
    "                    jh = j\n",
    "                if d_row[j] + D_ROW[j] + self.gapOpen > df:\n",
    "                    df = d_row[j] + D_ROW[j] + self.gapOpen\n",
    "                    jd = j\n",
    "                if s_row[j] + S_ROW[j] > st:\n",
    "                    st = s_row[j] + S_ROW[j]\n",
    "                    js = j\n",
    "\n",
    "            if max(hk, df, st) == st:\n",
    "                jmid = js\n",
    "                final_score = st\n",
    "                current = 'S'\n",
    "            \n",
    "            elif max(hk, df, st) == df:\n",
    "                jmid = jd\n",
    "                final_score = df\n",
    "                current = 'D'\n",
    "            \n",
    "            else:\n",
    "                jmid = jh \n",
    "                final_score = hk\n",
    "                current = 'H'\n",
    "            \n",
    "            # Recursively call itself on the partitions of both sequences. \n",
    "            top_align = self.gap3_trace(seq1[1: imid + 1], seq2[1:jmid + 1], s, d, h, \n",
    "                                         S_ROW[jmid], D_ROW[jmid], H_ROW[jmid], curr, mode)\n",
    "        \n",
    "            bottom_align = self.gap3_trace(seq1[imid+1::], seq2[jmid+1::], s_row[jmid], \n",
    "                                            d_row[jmid], h_row[jmid], S, D, H, current, mode)\n",
    "            \n",
    "            # Merge the two generated optimal alignments together to solve the optimal alignment of the \n",
    "            # whole sequences.\n",
    "            final_align_seq1 = top_align[1] + bottom_align[1]\n",
    "            final_align_mid = top_align[2] + bottom_align[2]\n",
    "            final_align_seq2 = top_align[3] + bottom_align[3]\n",
    "        \n",
    "            return [final_score, final_align_seq1, final_align_mid, final_align_seq2]\n",
    "         \n",
    "    # Method for gap3 alignment, which aligns both sequences end-to-end while finding multiple regions of \n",
    "    # similarities between the two sequences, and outputs the whole alignment. It is a generalized global \n",
    "    # alignment model defined to handle sequences with intermittent similarities. You can think of it as a\n",
    "    # mix of global alignment and local alignment. \n",
    "    # @param seq1 : first sequence\n",
    "    # @param seq2 : second sequence \n",
    "    # @param mode : determines the scoring system that the algorithm will use. Default is 0, which is the \n",
    "    #               scoring system for DNA/ RNA sequences. If not 0, the algorithm uses the scoring system \n",
    "    #               for protein sequences. \n",
    "    def gap3(self, seq1, seq2, mode=0):\n",
    "        s = 0\n",
    "        d = -self.gapOpen\n",
    "        h = -self.diff\n",
    "        if len(seq1) == 0 or len(seq2) == 0:\n",
    "            print(\"One of the sequences is empty.\")\n",
    "            return\n",
    "\n",
    "        arr = self.gap3_trace(seq1, seq2, s, d, h, s, d, h, 'S', mode)\n",
    "        self.printAlignment(seq1, seq2, 0, 0, len(seq1), len(seq2), arr[0], arr[1], arr[2], arr[3])\n",
    "    \n",
    "    # Method for formatting the optimal alignment output\n",
    "    # @param seq1 : first sequence\n",
    "    # @param seq2 : second sequence \n",
    "    # @param start1 : the index of the character in the first sequence that starts the generated alignment\n",
    "    # @param start2 : the index of the character in the second sequence that starts the generated alignment\n",
    "    # @param end1 : the index of the last character in the first sequence that ends the generated alignment\n",
    "    # @param end2 : the index of the last character in the second sequence that ends the generated alignment\n",
    "    # @param score : the alignment score of the optimal alignment\n",
    "    # @param align_seq1 : the alignment string consisting only the first sequence\n",
    "    # @param align_mid : the alignment string consisting only the middle symbols that define the matches ('|'),\n",
    "    #                    mismatches ('*'), difference blocks (\" \"), or same functionality amino acids ('='). \n",
    "    # @param align_seq2 : the alignment string consisting only the second sequence\n",
    "    def printAlignment(self, seq1, seq2, start1, start2, end1, end2, score, align_seq1, align_mid, align_seq2):\n",
    "        counter = 0\n",
    "        curr1 = start1 + 1\n",
    "        curr2 = start2 + 1\n",
    "        max_space = len(str(max(len(seq1), len(seq2))))\n",
    "        print(\"Alignment Score: \" + str(score))\n",
    "        print(\"Sequences Similarity: \" + str(round((align_mid.count(\"|\")/ len(align_mid)) * 100,2)) + \"%\\n\")\n",
    "        while counter <= len(align_seq1):\n",
    "            print(\"\\t\" + \"Sequence 1 > \" + self.generateString('left', max_space, str(curr1)) + align_seq1[counter: counter + 79] + self.generateString('right', max_space, str(min(curr1 + 79 - align_seq1[counter: counter + 79].count('-') - 1, end1))))\n",
    "            print(\"\\t\" + \"             \" + self.generateString('left', max_space, \"\") + align_mid[counter: counter + 79] + self.generateString('right', max_space, \"\"))\n",
    "            print(\"\\t\" + \"Sequence 2 > \" + self.generateString('left', max_space, str(curr2)) + align_seq2[counter: counter + 79] + self.generateString('right', max_space, str(min(curr2 + 79 - align_seq2[counter: counter + 79].count('-') - 1, end2))))\n",
    "            \n",
    "            print(\"\\n\")\n",
    "            curr1 = min(curr1 + 79 - align_seq1[counter: counter + 79].count('-'), end1)\n",
    "            curr2 = min(curr2 + 79 - align_seq2[counter: counter + 79].count('-'), end2)\n",
    "            counter += 79\n",
    "    \n",
    "    # Method for adding necessary number of spaces after and before numbers to make the output alignment\n",
    "    # cleaner and clearer to see. \n",
    "    # @param position : indicates the position of the numbers in the alignment. Add spaces to the right of the\n",
    "    #                   number if 'left', or left of the number if \"right\"\n",
    "    # @param max_space : the maximum number of digits the number has, allows for precise padding.\n",
    "    # @param string : the number in string format\n",
    "    def generateString(self, position, max_space, string):\n",
    "        if len(string) > max_space:\n",
    "            return\n",
    "        if position == 'left':\n",
    "            return string + ((max_space - len(string)) * \" \") + \" \"\n",
    "        elif position == 'right':\n",
    "            return \" \" + ((max_space - len(string)) * \" \") + string\n",
    "        else:\n",
    "            return\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "57c62f33",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GAP3 Alignment: \n",
      "Alignment Score: 760.0\n",
      "Sequences Similarity: 47.88%\n",
      "\n",
      "\tSequence 1 > 1   GCGCTCCGGGACGCCTTCCGCCGTCGGGAGCCCTACAACTACCTGCAGAGGGCCTATTACCAGGTGGGGAGCGGGCCGG  79\n",
      "\t                 +++++++++++++++++++++++++|||||||*|||||||||||||||||||||||*|||||||||*||+++|||||||    \n",
      "\tSequence 2 > 1                            GGGAGCCTTACAACTACCTGCAGAGGGCCTACTACCAGGTGCGG   GGGCCGG  79\n",
      "\n",
      "\n",
      "\tSequence 1 > 80  GCAGTAGC    CTTCCCCAGAGCCCCCTAGCCGCA              GGCAC     CAGAGGGTCCAAGACAAGA 158\n",
      "\t                 *|||++++++++||*|||||+||+++||||++++++++++++++++++++||||++++++|||*|*|||||*||*||++    \n",
      "\tSequence 2 > 80  CCAG    GGTGCTACCCCA AG   CCTA      CTGACTGTCTTACTGGCA AGCTTCAGCGAGTCCAGGAGAA   158\n",
      "\n",
      "\n",
      "\tSequence 1 > 159 CT     GGAAGGGCACCT  CGGGTTCGG      GAGGAGCTGTGAGTGGCT                         176\n",
      "\t                 +++++++||||++||*||+++|||||*|||++++++|||||*||||||*|||||++++++++++++++++++++++++    \n",
      "\tSequence 2 > 159   AGCTGGGAA  GCCCC GCCGGGTCCGGGTCCGAGAGGAACTGTGAATGGCTGAGCCTGCTTCTCGAGGATCAGGC 186\n",
      "\n",
      "\n",
      "-----------------------------------------------------------------------------------------------------\n",
      "Global Alignment: \n",
      "Alignment Score: 454.0\n",
      "Sequences Similarity: 53.49%\n",
      "\n",
      "\tSequence 1 > 1   GCGCTCCGGGACGCCTTCCGCCGTCGGGAGCCCTACAACTACCTGCAGAGGGCCTATTACCAGGTGGGGAGCGGGCCGG  79\n",
      "\t                                          |||||||*|||||||||||||||||||||||*|||||||||*||   |||||||    \n",
      "\tSequence 2 > 1   -------------------------GGGAGCCTTACAACTACCTGCAGAGGGCCTACTACCAGGTGCGG---GGGCCGG  51\n",
      "\n",
      "\n",
      "\tSequence 1 > 80  GCAGTAGCCTTCCCCAGAGCC---------CCCTAGCCGCAGGCACCAGAGGGTCCAAGACAAGACTGGAAGGGCACCT 149\n",
      "\t                 *|||****||*||||| ||||         *|*||***|||*||**|||*|*|||||*||*||**||||*|*|*|*|**    \n",
      "\tSequence 2 > 52  CCAGGGTGCTACCCCA-AGCCTACTGACTGTCTTACTGGCAAGCTTCAGCGAGTCCAGGAGAAAGCTGGGAAGCCCCGC 129\n",
      "\n",
      "\n",
      "\tSequence 1 > 150 CGGGTTCGG------GAGGAGCTGTGAGTGGCT------------------------ 176\n",
      "\t                 |||||*|||      |||||*||||||*|||||                            \n",
      "\tSequence 2 > 130 CGGGTCCGGGTCCGAGAGGAACTGTGAATGGCTGAGCCTGCTTCTCGAGGATCAGGC 186\n",
      "\n",
      "\n",
      "-----------------------------------------------------------------------------------------------------\n",
      "Local Alignment: \n",
      "Alignment Score: 632.0\n",
      "Sequences Similarity: 69.28%\n",
      "\n",
      "\tSequence 1 > 26  GGGAGCCCTACAACTACCTGCAGAGGGCCTATTACCAGGTGGGGAGCGGGCCGGGCAGTAGCCTTCCCCAGAGCC---- 100\n",
      "\t                 |||||||*|||||||||||||||||||||||*|||||||||*||   |||||||*|||****||*||||| ||||        \n",
      "\tSequence 2 > 1   GGGAGCCTTACAACTACCTGCAGAGGGCCTACTACCAGGTGCGG---GGGCCGGCCAGGGTGCTACCCCA-AGCCTACT  75\n",
      "\n",
      "\n",
      "\tSequence 1 > 101 -----CCCTAGCCGCAGGCACCAGAGGGTCCAAGACAAGACTGGAAGGGCACCTCGGGTTCGG------GAGGAGCTGT 168\n",
      "\t                      *|*||***|||*||**|||*|*|||||*||*||**||||*|*|*|*|**|||||*|||      |||||*||||    \n",
      "\tSequence 2 > 76  GACTGTCTTACTGGCAAGCTTCAGCGAGTCCAGGAGAAAGCTGGGAAGCCCCGCCGGGTCCGGGTCCGAGAGGAACTGT 154\n",
      "\n",
      "\n",
      "\tSequence 1 > 169 GAGTGGCT 176\n",
      "\t                 ||*|||||    \n",
      "\tSequence 2 > 155 GAATGGCT 162\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "p = Sequence_Alignment(mismatch=-10, gapOpen=40, gapExtend=2, diff_block=20)\n",
    "A = \"GCGCTCCGGGACGCCTTCCGCCGTCGGGAGCCCTACAACTACCTGCAGAGGGCCTATTACCAGGTGGGGAGCGGGCCGGGCAGTAGCCTTCCC\"\n",
    "A+=\"CAGAGCCCCCTAGCCGCAGGCACCAGAGGGTCCAAGACAAGACTGGAAGGGCACCTCGGGTTCGGGAGGAGCTGTGAGTGGCT\"\n",
    "B = \"GGGAGCCTTACAACTACCTGCAGAGGGCCTACTACCAGGTGCGGGGGCCGGCCAGGGTGCTACCCCAAGCCTACTGACTGTCTTACTGG\"\n",
    "B+= \"CAAGCTTCAGCGAGTCCAGGAGAAAGCTGGGAAGCCCCGCCGGGTCCGGGTCCGAGAGGAACTGTGAATGGCTGAGCCTGCTTCTCGAGGATCAGGC\"\n",
    "print(\"GAP3 Alignment: \")\n",
    "p.gap3(A,B)\n",
    "print(\"-----------------------------------------------------------------------------------------------------\")\n",
    "print(\"Global Alignment: \")\n",
    "p.global_alignment(A,B)\n",
    "print(\"-----------------------------------------------------------------------------------------------------\")\n",
    "print(\"Local Alignment: \")\n",
    "p.local_alignment(A,B)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "b87028ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment Score: 354.0\n",
      "Sequences Similarity: 26.43%\n",
      "\n",
      "\tSequence 1 > 1   GCGCTCCGGGACGCCTTCCGCCGTCGGGAGCCCTACAACTACCTGCAGAGGGCCTATTACCAGGTGGGGAGCGGGCCGG  79\n",
      "\t                 +++++++++++++++++++++++++|||||||*|||||||||||||||||||||||*|||||||||*||   |||||||    \n",
      "\tSequence 2 > 1                            GGGAGCCTTACAACTACCTGCAGAGGGCCTACTACCAGGTGCGG---GGGCCGG  76\n",
      "\n",
      "\n",
      "\tSequence 1 > 80  GCAGTAGCCTTCCCCAGAGCCCCCTAGCCGCAGGCACCAGAGGGTCCAAGACAAGACTGGAAGGGCACCT          158\n",
      "\t                 *|||+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++    \n",
      "\tSequence 2 > 77  CCAG                                                                  GGTGCTACC 155\n",
      "\n",
      "\n",
      "\tSequence 1 > 159                                                                        CGGGTTCG 176\n",
      "\t                 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++|||||*||    \n",
      "\tSequence 2 > 156 CCAAGCCTACTGACTGTCTTACTGGCAAGCTTCAGCGAGTCCAGGAGAAAGCTGGGAAGCCCCGCCGGGTCCGGGTCCG 186\n",
      "\n",
      "\n",
      "\tSequence 1 > 176 GGAGGAGCTGTGAGTGGCT                         176\n",
      "\t                 *|||||*||||||*|||||++++++++++++++++++++++++    \n",
      "\tSequence 2 > 186 AGAGGAACTGTGAATGGCTGAGCCTGCTTCTCGAGGATCAGGC 186\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "p = Sequence_Alignment(diff_block=60)\n",
    "A = \"GCGCTCCGGGACGCCTTCCGCCGTCGGGAGCCCTACAACTACCTGCAGAGGGCCTATTACCAGGTGGGGAGCGGGCCGGGCAGTAGCCTTCCC\"\n",
    "A+=\"CAGAGCCCCCTAGCCGCAGGCACCAGAGGGTCCAAGACAAGACTGGAAGGGCACCTCGGGTTCGGGAGGAGCTGTGAGTGGCT\"\n",
    "B = \"GGGAGCCTTACAACTACCTGCAGAGGGCCTACTACCAGGTGCGGGGGCCGGCCAGGGTGCTACCCCAAGCCTACTGACTGTCTTACTGG\"\n",
    "B+= \"CAAGCTTCAGCGAGTCCAGGAGAAAGCTGGGAAGCCCCGCCGGGTCCGGGTCCGAGAGGAACTGTGAATGGCTGAGCCTGCTTCTCGAGGATCAGGC\"\n",
    "p.gap3(A,B)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "88857e0c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Aligning NW_021636346.1:c40256-27400 with NW_021636346.1:c40256-27400 using gap3 alignment: \n",
      "\n",
      "Alignment Score: 22500.0\n",
      "Sequences Similarity: 33.33%\n",
      "\n",
      "\tSequence 1 > 1    TCACGGAGCTTAAACAAGGCCCTCTATAGCCTCTGTCTGCTTACTGTGCCACAAAAACGCCCACCTTCATAACGGCTTC   79\n",
      "\t                  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 1    TCACGGAGCTTAAACAAGGCCCTCTATAGCCTCTGTCTGCTTACTGTGCCACAAAAACGCCCACCTTCATAACGGCTTC   79\n",
      "\n",
      "\n",
      "\tSequence 1 > 80   TTGCATAAGCGACTGTGGAAGAAAGCTGGGGAGAGTGGTGGAGTTGATTGCCATGGGGTTTTCGACCCAGAAGCCCACC  158\n",
      "\t                  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 80   TTGCATAAGCGACTGTGGAAGAAAGCTGGGGAGAGTGGTGGAGTTGATTGCCATGGGGTTTTCGACCCAGAAGCCCACC  158\n",
      "\n",
      "\n",
      "\tSequence 1 > 159  ACCACCACTTTGCGGCCCTTTTGCCCCTCCTCTTTCTGGACTTCGCGCTCTTGTCGTTCCTTTCGCGTCGCTTCTTCTT  237\n",
      "\t                  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 159  ACCACCACTTTGCGGCCCTTTTGCCCCTCCTCTTTCTGGACTTCGCGCTCTTGTCGTTCCTTTCGCGTCGCTTCTTCTT  237\n",
      "\n",
      "\n",
      "\tSequence 1 > 238  CTTGGCCTCTCATGGCTCCCTCCGAAGCTCACTATGGCAGGTTCACTTCTGCTGTTCTTTTTTTACGTTCTCTGCAAGT  316\n",
      "\t                  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 238  CTTGGCCTCTCATGGCTCCCTCCGAAGCTCACTATGGCAGGTTCACTTCTGCTGTTCTTTTTTTACGTTCTCTGCAAGT  316\n",
      "\n",
      "\n",
      "\tSequence 1 > 317  TTTTACCTCTTGGCGCATGGGTTACTGTTTTTCACTGCAGATTTTTGAAATTTAATTCAATATCCTTACGTTTTAGGGA  395\n",
      "\t                  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 317  TTTTACCTCTTGGCGCATGGGTTACTGTTTTTCACTGCAGATTTTTGAAATTTAATTCAATATCCTTACGTTTTAGGGA  395\n",
      "\n",
      "\n",
      "\tSequence 1 > 396  GGGCATAAGTGTTTTCAGGGAAACATGGCCTTGTCATGGTAAATGTAATGGTCAATTTTTTTTTTTTGGTGAATTCACT  474\n",
      "\t                  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 396  GGGCATAAGTGTTTTCAGGGAAACATGGCCTTGTCATGGTAAATGTAATGGTCAATTTTTTTTTTTTGGTGAATTCACT  474\n",
      "\n",
      "\n",
      "\tSequence 1 > 475  CTATCTATTTATAAGTGCAGTTGTCATTGTTTTGTTGGTAATGCACTTTTCCTAGATGTTTCCCAAAAAAGGTAGTTGG  553\n",
      "\t                  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 475  CTATCTATTTATAAGTGCAGTTGTCATTGTTTTGTTGGTAATGCACTTTTCCTAGATGTTTCCCAAAAAAGGTAGTTGG  553\n",
      "\n",
      "\n",
      "\tSequence 1 > 554  TAGCACATTGTCGGAGCTGAACGTTGACGTTTTGTTTTTCGGTGCACATTTTCGAAATTCGTTAGATAATCCAAAGTTT  632\n",
      "\t                  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 554  TAGCACATTGTCGGAGCTGAACGTTGACGTTTTGTTTTTCGGTGCACATTTTCGAAATTCGTTAGATAATCCAAAGTTT  632\n",
      "\n",
      "\n",
      "\tSequence 1 > 633  TAGACGAATGTTTTGGCGAAAGCCCATACTAACCAAATGGTCAGTTCGTTTGGCCTATTTGTTTATATTATTCAATGAC  711\n",
      "\t                  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 633  TAGACGAATGTTTTGGCGAAAGCCCATACTAACCAAATGGTCAGTTCGTTTGGCCTATTTGTTTATATTATTCAATGAC  711\n",
      "\n",
      "\n",
      "\tSequence 1 > 712  CTCTTTGTGCATTCAAAATTTTTGTAACGCATTTGTTGTTGTTTGTGGGGGTGGGGAGGGGGCTTTGAATTTTGAGAAG  790\n",
      "\t                  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 712  CTCTTTGTGCATTCAAAATTTTTGTAACGCATTTGTTGTTGTTTGTGGGGGTGGGGAGGGGGCTTTGAATTTTGAGAAG  790\n",
      "\n",
      "\n",
      "\tSequence 1 > 791  TTAGTGTAGTAGCATATGTAATAAGTCAATGGTGTCATCTCGAGGACTTTTTCTGGCTAGCTTAAAATCTTGTTTGGCC  869\n",
      "\t                  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 791  TTAGTGTAGTAGCATATGTAATAAGTCAATGGTGTCATCTCGAGGACTTTTTCTGGCTAGCTTAAAATCTTGTTTGGCC  869\n",
      "\n",
      "\n",
      "\tSequence 1 > 870  TAAGGTGATTATAAACGTCTTGCTTCTTCTTCTGTGGGTTAAATTGATATTTAAATTCATTATTGTGTCTAAGTGGTTG  948\n",
      "\t                  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 870  TAAGGTGATTATAAACGTCTTGCTTCTTCTTCTGTGGGTTAAATTGATATTTAAATTCATTATTGTGTCTAAGTGGTTG  948\n",
      "\n",
      "\n",
      "\tSequence 1 > 949  CGACTTTTATTTTCTTTGCATGATATGAACAGCAGAAGAAAAATCCTCTCCGTTCAGGTAGTCAACAATGGCTGGCTGA 1027\n",
      "\t                  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 949  CGACTTTTATTTTCTTTGCATGATATGAACAGCAGAAGAAAAATCCTCTCCGTTCAGGTAGTCAACAATGGCTGGCTGA 1027\n",
      "\n",
      "\n",
      "\tSequence 1 > 1028 ATAGCACTGTGTCTTGTAATTCTCTTCAATTTAGAGAAGGTTCCTTACAATTTCTATCAAGAATAGTTATTGGGATGAA 1106\n",
      "\t                  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 1028 ATAGCACTGTGTCTTGTAATTCTCTTCAATTTAGAGAAGGTTCCTTACAATTTCTATCAAGAATAGTTATTGGGATGAA 1106\n",
      "\n",
      "\n",
      "\tSequence 1 > 1107 TAAACACAATGATATGGAATATAACCGTCATCTGTATAAGACTCCAATTCATGGTTCATTCATGAAGTTTTCGACTGCA 1185\n",
      "\t                  ||||||||||||||+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 1107 TAAACACAATGATA                                                                  1185\n",
      "\n",
      "\n",
      "\tSequence 1 > 1186 TCGACTGGCATGTTGAAGGAAGTGACAGATGAAGAAGAGCTCAATTCTTCACATTATTCTGAAGCAAAAGTTCAGACTG 1264\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 1186                                                                                 1264\n",
      "\n",
      "\n",
      "\tSequence 1 > 1265 TAAGTATTTTTGCTCATGACTGTGCAACATCCAATATTGCTGCTAAAGGTGAACTGTTTGGTGATAGAAGAGGCTGGGT 1343\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 1265                                                                                 1343\n",
      "\n",
      "\n",
      "\tSequence 1 > 1344 GCGAGCTAGAGATAATTTGACTGTAACTAAAATGCAAAATAAGTTTGGGTGCAATAATGGCAGCACTTATGCATCTAAA 1422\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 1344                                                                                 1422\n",
      "\n",
      "\n",
      "\tSequence 1 > 1423 ATATGTATTTCCAATGGCCAGGAGCAGGTTCTTTCTAAGGACGGAGGGCCCCGTCCTCTTTTACTGAAGGACCATGAGG 1501\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 1423                                                                                 1501\n",
      "\n",
      "\n",
      "\tSequence 1 > 1502 TACCTATCTCATCCCCTGTGGCCAGCCATTCTAAATCGAAAAATACAAAGGTACTGATGGACTCTAAAAGGGCAGGCGA 1580\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 1502                                                                                 1580\n",
      "\n",
      "\n",
      "\tSequence 1 > 1581 GTCAATTGAATGCCATTTGGTAATCAATAAAGATGAAACTTCTGATATAAATGAGAATCACAATTTGGAGGCATCTACC 1659\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 1581                                                                                 1659\n",
      "\n",
      "\n",
      "\tSequence 1 > 1660 AAGCATGCTGACATTGCATTAATGTCCAAAACAAAATATGTTACTGGTGATTCAAAGCTTCGCGCCAGGCTCTGTGGCA 1738\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 1660                                                                                 1738\n",
      "\n",
      "\n",
      "\tSequence 1 > 1739 TTTATAATGATATATTGGTCGTTGATAATATTTCTCTAGCTAAGGGAGTTGTTGGAATGCTTACAAAAAAGTACAGGCA 1817\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 1739                                                                                 1817\n",
      "\n",
      "\n",
      "\tSequence 1 > 1818 TCTTGTTCATGCATGTGATACTGAGGTATAAAGTCTTTCTAAGCATGCTAATTTATGTTGTGGTTTTCATTGGTACAAC 1896\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 1818                                                                                 1896\n",
      "\n",
      "\n",
      "\tSequence 1 > 1897 ATTTTCTATAGCTTTGTTATTTATTTTCTTTTACTTTGGGCTTGAATTTGACCCGTTTGACCATTTCAGCTTGTGTATA 1975\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 1897                                                                                 1975\n",
      "\n",
      "\n",
      "\tSequence 1 > 1976 ATATAAAATCTTTCTTTCTTAAAATTCTTTTCTTCTTCTATGCTTTTATTAACATTATGGTCAAATGTTCTGAGATGAG 2054\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 1976                                                                                 2054\n",
      "\n",
      "\n",
      "\tSequence 1 > 2055 GCAATAAGAGGCCTTTGTGCCTATTGCCATTTCTAAGCGTGATTCCATGGCGTCAAAAGTGATTTTTCATACATCTTTT 2133\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2055                                                                                 2133\n",
      "\n",
      "\n",
      "\tSequence 1 > 2134 GACTATTTTTGTTTTGCAAATTGTGGACATATAATATGTGCTGGTGAAAACATCTTTAAAGGTGGCCAAGATTGATGTC 2212\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2134                                                                                 2212\n",
      "\n",
      "\n",
      "\tSequence 1 > 2213 AAAGAGGAAACACCTGTTGATCATGGGGAGATAGTATGTTTCAGTGTTTATTCTGGTCCAGAAGCTGATTTTGGAAATG 2291\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2213                                                                                 2291\n",
      "\n",
      "\n",
      "\tSequence 1 > 2292 GGAAATCCTGTATTTGGGTGGATGTTCTTGATGGTGGAGGCAAAGAGATGTTATTTGAATTTGAACCGTTTTTCAAAGA 2370\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2292                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 2371 CCAATCCATCAAAAAGGTGAATTTTCCTGCTTAGCATCCTATTTTTGTTTGTTCTTGTTAAATTTTAATGTTTAGCTGC 2449\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 2450 TTATATGAAATTAATGGTAAATATTCTAGATTGGGGGGAGGATATTAAAAGAGATACACCTGAAAAAATAACTCTTGAG 2528\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 2529 CTAATGGTAGCGCATTACTCGCAGTCTTTCATGTGAGTTGGAAAATTAATTTGAAATAACAAAGCCACCCGTATCTTCT 2607\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 2608 CCAGGTCTGGCATAACTATAGCTTTGATAATCATGTTATAGAGAACTATGGGCTTAAAGTTTCTGGTTTTCATGCTGAT 2686\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 2687 ACAATGCACATGGCACGATTATGGAATTCATCAAGACGCTTGGATGGTGGTTATTCTCTTGAAGCACTTACAAGTGATA 2765\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 2766 AAAGGGTTATGTGTGGGGCTCATTTGAACCATGAAAAGGATTTGATTGGTAAAGTGTCGATGAAAACTATATTTGGGCG 2844\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 2845 GAAAAAGGTGAAAAAAGATGGATCTGAGGGTAAAATGATTACTATTCCTCCTGTTGAAGAGCTACAAAGAGATGAACGA 2923\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 2924 ATACCTTGGATATGTTATTCTGCCTTAGATGCTAGGAGCACTTTGAAGCTTTATGAAAGCCTTAAGAGCCATCTCTCAA 3002\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 3003 AAATGCTGTGGAAAGTTGATGGAATACAGATGATGGGACTGGTGTCTGCAAAAACCATGTATGATTTCTATGAAGAATA 3081\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 3082 TTGGCGGCCATTTGGTGAACTTCTAGTCCAAATGGAATCTGAAGGGATGCTAGTTGACCGTGCATATCTTGCTGAGATA 3160\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 3161 GAGAAAGTGGCCAAGGCAGAGCAAGAGGTAGCTGTTAACAGGTTCCGGAAATGGGCGTCTAGGTATTGTCCTGATGCAA 3239\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 3240 AGTATATGAATGTGGGAAGTGATGCTCAGTTGCGCCAGCTGCTATTTGGTGGCATGGTGAACAGGTGCGCATCTATTTA 3318\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 2310                                                    GCATGGTGAACAGGTGCGCATCTATTTA 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 3319 TTCTTCTCTGCTTCTTTTGATTCTTCCTTATTTTGTAAAATGAAGACTTAAATTTGTTTACATATAACCTATATTCAAT 3397\n",
      "\t                  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 2310 TTCTTCTCTGCTTCTTTTGATTCTTCCTTATTTTGTAAAATGAAGACTTAAATTTGTTTACATATAACCTATATTCAAT 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 3398 ATACAGAAGACATGTCTCTATTATGATGCTATTGTCATAAAATACGAGGTCCACTATTTATTAGATGCATACAAATATC 3476\n",
      "\t                  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 2310 ATACAGAAGACATGTCTCTATTATGATGCTATTGTCATAAAATACGAGGTCCACTATTTATTAGATGCATACAAATATC 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 3477 TAAGCTAAGTTCGTCTCTAGTTTGAAAGCCTCTGTTATTGCAAATAGGCTTTTTTATTATTATTACAATTGTTTGCTAC 3555\n",
      "\t                  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 2310 TAAGCTAAGTTCGTCTCTAGTTTGAAAGCCTCTGTTATTGCAAATAGGCTTTTTTATTATTATTACAATTGTTTGCTAC 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 3556 AGTAATCGTTTTCTTGAAAATTTTAATTTGGATTCTAAATGCTTTCAAGTGTAGAATATATTTCTTCACTGGAATTAAT 3634\n",
      "\t                  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 2310 AGTAATCGTTTTCTTGAAAATTTTAATTTGGATTCTAAATGCTTTCAAGTGTAGAATATATTTCTTCACTGGAATTAAT 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 3635 TTGACTTAATAGATTAACCGGACACCAAATCTCTGTGAATTATGATTTAGAGTCTGGACTTGTTCATCCGAAATGGTAC 3713\n",
      "\t                  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 2310 TTGACTTAATAGATTAACCGGACACCAAATCTCTGTGAATTATGATTTAGAGTCTGGACTTGTTCATCCGAAATGGTAC 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 3714 TCATTGCATTAGTTTGCAGGGCCGTAAGATTAGTACATAGTAGAAACTCAGTTTTAAAGTAGGTATTGATATTTTCAAT 3792\n",
      "\t                  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 2310 TCATTGCATTAGTTTGCAGGGCCGTAAGATTAGTACATAGTAGAAACTCAGTTTTAAAGTAGGTATTGATATTTTCAAT 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 3793 GTTAATCATAGTGCAACTCAACTGGGTTAGTCTGAGGCTTGATGTATTCCAGAATCCAAACACAAGCTGCTGTTGGTTC 3871\n",
      "\t                  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 2310 GTTAATCATAGTGCAACTCAACTGGGTTAGTCTGAGGCTTGATGTATTCCAGAATCCAAACACAAGCTGCTGTTGGTTC 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 3872 GGGTTTTGGGCCAATCACAAACAAACAACATGGGTCATATGTTTATGAACTTAGACCCTTTAATATATGTACATATTGT 3950\n",
      "\t                  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 2310 GGGTTTTGGGCCAATCACAAACAAACAACATGGGTCATATGTTTATGAACTTAGACCCTTTAATATATGTACATATTGT 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 3951 TGTACATGAGATCATGGGAAAAAAGAAGTGCTTTTCTTTCTCATTTCTAATTAAACTTACTATTTTCCACTATTCACAA 4029\n",
      "\t                  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 2310 TGTACATGAGATCATGGGAAAAAAGAAGTGCTTTTCTTTCTCATTTCTAATTAAACTTACTATTTTCCACTATTCACAA 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 4030 AAATGTAGTTCAGCTACAGTGATAAATATGAGGCTTTCATTTTCATTTTCTGAATGCTTATTCCAGTAAAGACCCTAGT 4108\n",
      "\t                  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 2310 AAATGTAGTTCAGCTACAGTGATAAATATGAGGCTTTCATTTTCATTTTCTGAATGCTTATTCCAGTAAAGACCCTAGT 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 4109 GTGGCACTTCCAATAGAGCGGACTTTCAAAGTTCCCAATGTTGATAAAGTCATCAAAGAAGGCCAGAAGGCTCCCTCAA 4187\n",
      "\t                  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 2310 GTGGCACTTCCAATAGAGCGGACTTTCAAAGTTCCCAATGTTGATAAAGTCATCAAAGAAGGCCAGAAGGCTCCCTCAA 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 4188 AATTTCGCGACATTAAGCTGCATAGCATTGGTTGTCAACTGGAAACTGAGATGTACACAGAATCTGGTTGGCCATCTGT 4266\n",
      "\t                  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 2310 AATTTCGCGACATTAAGCTGCATAGCATTGGTTGTCAACTGGAAACTGAGATGTACACAGAATCTGGTTGGCCATCTGT 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 4267 CAGTGGTGATGCTTTAAAGGCTCTTGCTGGAAATGCTTCAGCTGAATATGACTTTCTAGATGAGGCCTGTAACCTGGGT 4345\n",
      "\t                  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 2310 CAGTGGTGATGCTTTAAAGGCTCTTGCTGGAAATGCTTCAGCTGAATATGACTTTCTAGATGAGGCCTGTAACCTGGGT 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 4346 TGTGATGATGAATATGAAAATTCTTCTAAGGATGAAGTTACCTATGTGGAAAAGGACACTACAGCTTATGGAACTGCTT 4424\n",
      "\t                  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||     \n",
      "\tSequence 2 > 2310 TGTGATGATGAATATGAAAATTCTTCTAAGGATGAAGTTACCTATGTGGAAAAGGACACTACAGCTTATGGAACTGCTT 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 4425 TTTCTGCCTTTGCAACTGAGGATGAAGGTAGAGAAGCTTGCCATGCAATTGCTGCTTTATGCGAAGTCTGTTCAATCGA 4503\n",
      "\t                  ||||||||||||||||||||||||||||||||||||||||||||||||||||||||+++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310 TTTCTGCCTTTGCAACTGAGGATGAAGGTAGAGAAGCTTGCCATGCAATTGCTGCT                        2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 4504 CTCTTTGATATCTAACTTTATCCTTCCCTTGCAGGTAATTTATTCTATTTGTATTTCTGTTAGATTTTCCTCCCTTTTG 4582\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 4583 TTTATGCTAATCATTCATTGTTGCGTTCCCCTGATAATCTTTGACTACTTTGGCTGAGGCATATTTTTAATGAAGTTTT 4661\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 4662 GGTTGGAAACTTTTTAAGTTTCAACATTATTTGAGCATGAGTGACAGGTGGGGGAGGGTAGCTTGCTATCTATTTACTC 4740\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 4741 GTCCACTAATGCAATTTTCTTTATTTTATTCTAGGATTGTTAATATTTTTCTAGTTTTCTTGGTTTTGGTCTGTGGCAG 4819\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 4820 GGTAGTAATGTTTCGGGAAGGAATGGGCGCATTCATTGTTCCTTAAATATCAACACGGAGACAGGACGCTTGTCAGCTA 4898\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 4899 GAAGGCCTAATTTACAGGTTCTTTTAAATTAAAAGACATGATTAAAGTCATCTTTATGATTATGATTAGGTTACTTTTG 4977\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 4978 AATGTTAAATTTTGGTTTACTTGAAAAAGCTCGAGTACTGTACAATATAAACTCTACCATTGGCCTCCGTTTTTGGTAT 5056\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 5057 ATGGGCATGTTGTTCCTAAATTAGAAGTTGGATCACACATTGTAGCAGTGCACAAAAAAAGGGACTGGACAGTTCGGCT 5135\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 5136 TAGACCAGAAACCGGCCTCCGATCAGCTCGGAATAAACCGGGATGAACGAGGAAAAACCGAGTAACCCACTGGTTCGTG 5214\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 5215 CAAGTTGGCCGGTTTCTTTTTTTTTTTTTAAAAAAAAAAGCTTCTTTCTTGCTCACTGTTTCCCGGAGGAGCCATGGAA 5293\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 5294 AGTGGAAACTACTCGTGCTCTGCCTCTGCTCGTCGAAGTCAAGGAGGAAGAAGTGTAATGAGTGACTGATGTGTGTTCA 5372\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 5373 AGGAGGAAAGAAGCTGTGATGGATTGATGCATACTGAAGGAGGAAAGAGAGTCTGCACTCTACATGTTCAAGGGGGGAA 5451\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 5452 GATGAAAGAGGAAAGAAGTTGTGTATGCCATGTGATGCATGTGTATTAGGGAGAGGAGGAAAGGTCAAAAGGTAGAGTG 5530\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 5531 CTAATGATCACGTGGGAGAGAGAACAAAAATGAGGAAAGAAGGAACAAAGGGAGTTAAAAAAAAAAAAAAATACAAAGA 5609\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 5610 GGAGATAATTTATTGAGTCACTGGCGGTAGAAAGAGGAAAATAAAAGAAAGAAAAAAAAATTCATATGTATAAATACTT 5688\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 5689 TATAACATAAGTAATTATTTAAAAATTAAAAATTATTTATTATAATTAAAATATGAATATTTTAACTCATTTTAAATAA 5767\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 5768 ATTCTAACATTTTAATATAACTGGTTTGACTCTGAGTTGATCCCGATATGACTGTTTAACCCTCAAACCACTGTGTTAA 5846\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 5847 CTGGTTTAGTATCTGGTCCAGTTTTTTATTCCTTGCATTGCAGTATTATTTCACAATATACTGTACTACTTGATGCTAG 5925\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 5926 TATGTGGGTTTTACTTACTACACAACTCTATCTGTAATAGTAGTTGCATAGCAACCCATTTTCTATCTTTCTTATGAAT 6004\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 6005 GGCTGCAACAATATTTTATGGCACTAACTTATTATTGTTGTTGTTGTTGTTGTTATTTTTAAGATTATGTCCTTTTAAC 6083\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 6084 TTTGTCTCTCATTTTGTTATGATGCATCATAGAATGCTTTATCTTCTTTGCTTATCTTTATCTTTATTCAAATAAAAAC 6162\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 6163 GTATCCAAACTTTAATTCATCATTACATTTTGGTGAGGAAATTTATATGGTTATGATTTTCCTTAACTAAAGTGAATGA 6241\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 6242 ACCTTTCTTCTCTCTTTTTTTCAATTTTTTTTTGTTCTTTGTATTAACAGCAGCTTTATTATGGATGACATGGGATCCT 6320\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 6321 GTTGGGCATACAGGGGCAGGGCTGAAATTCTGTTTTGCTTGTTAAGTATTTGAAGTTTCCATTGGAAGCATCTTCTCTC 6399\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 6400 TTTTCTGGACGTCGGATGTCTACTTGGTCTTTTGTTTCATTAGCTATTAATTCGATAGAGTTGATGGATACTGTTAGGT 6478\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 6479 GGTTTCATTTGTTTTGGGGTTTTCAGTCGGACTTTCACGTGTAGGTATCACATTGACTAGATTTATAATTGTTATTCAA 6557\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 6558 AATCTTTAGAGCTCATCTAATATATGTGTGCTTTTAATAAAATAGAATCAACCTGCGTTGGAAAAGGATCGGTACAAGA 6636\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 6637 TTCGTCAAGCTTTTATAGCTGCACCTGGTAATTCTCTTATAGTTGCTGATTATGGACAGGTTAGCAGTTTATGCTATTG 6715\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 6716 ATTTATTTGCGTTGATTGCAATAAAGTTGTTCTGTTGATTGAGATGTTTATTGTGTTACAGCTGGAGCTAAGAATTCTT 6794\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 6795 GCGCACCTTGCCAGCTGTAAGAGTATGCTGGATGCCTTTAAAGCTGGTGGAGATTTCCATTCAAGGACTGCAATGAATA 6873\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                                                 2310\n",
      "\n",
      "\n",
      "\tSequence 1 > 6874 TGTATCCACATATTCGAGAAGCAATTGAAAAAAAGGAAGTGGTTCTGGAGTGGCATC 6930\n",
      "\t                  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++     \n",
      "\tSequence 2 > 2310                                                           2310\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "p = Sequence_Alignment(gapExtend=2)\n",
    "p.alignInput(\"gene.fna\", \"mutant.fna\", \"gap3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "7268589d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter a PROTEIN sequence to be aligned: gQQTA\n",
      "------------------------------------------------------------------------------------------------\n",
      "Enter another PROTEIN sequence to be aligned: ooa\n",
      "***PROTEINsequences should contain only these characters: AGVLISTCMPFYWDENQKRH, and no spaces are allowed ! ***\n",
      "\n",
      "Enter another PROTEIN sequence to be aligned: AQTYHC\n",
      "------------------------------------------------------------------------------------------------\n",
      "Please choose an alignment mode: \n",
      "\t1. Type 'global' for global alignment\n",
      "\t2. Type 'local' for local alignment\n",
      "\t3. Type 'gap3' for gap3 alignment\n",
      "\n",
      "Enter your desired alignment type: global\n",
      "------------------------------------------------------------------------------------------------\n",
      "\n",
      "Outputting global Alignment ...\n",
      "\n",
      "Alignment Score: -40.0\n",
      "Sequences Similarity: 16.67%\n",
      "\n",
      "\tSequence 1 > 1 gQ-QTA 5\n",
      "\t               *| ***  \n",
      "\tSequence 2 > 1 AQTYHC 6\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "p = Sequence_Alignment()\n",
    "p.manualAlign('protein')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "ffdb1442",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GAP3 Alignment: \n",
      "Alignment Score: -20.0\n",
      "Sequences Similarity: 0.0%\n",
      "\n",
      "\tSequence 1 > 1   RFQLVEQLHWCGYNGEVTGCGQSQKMHRKPFNRRQVYNKAHMISLGVQPL                               79\n",
      "\t                 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++    \n",
      "\tSequence 2 > 1                                                     FTQQDEFVWLEWNDLRVKQEIEVDSAYEP  79\n",
      "\n",
      "\n",
      "\tSequence 1 > 80                                     QFQYRFTINYKSKWIVCNFDVPFWVENELPRADTRWAYSMIQKD 101\n",
      "\t                 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++    \n",
      "\tSequence 2 > 80  MIWKECKANSYLNWQWMRGIEDHVADFVIFHHQNW                                             101\n",
      "\n",
      "\n",
      "\tSequence 1 > 101 VGHQAAV                                      101\n",
      "\t                 ++++++++++++++++++++++++++++++++++++++++++++    \n",
      "\tSequence 2 > 101        QFVMIVIQMWAEMFEDDSICATARCWIYNRFLWSHRP 101\n",
      "\n",
      "\n",
      "-----------------------------------------------------------------------------------------------------\n",
      "Global Alignment: \n",
      "Alignment Score: -472.0\n",
      "Sequences Similarity: 1.55%\n",
      "\n",
      "\tSequence 1 > 1   RFQLVEQLHWCGYNGEVTGCGQSQKMHRKPFNRRQVYNKAHMISLGVQPLQFQYRFTINYKSKWIVCNFDVPFWVENEL  79\n",
      "\t                 **|**|                                                                             \n",
      "\tSequence 2 > 1   FTQQDE-------------------------------------------------------------------------   6\n",
      "\n",
      "\n",
      "\tSequence 1 > 80  PRADTRWAYSMIQKDVGHQAAV--------------------------------------------------------- 101\n",
      "\t                                     *|                                                             \n",
      "\tSequence 2 > 7   --------------------FVWLEWNDLRVKQEIEVDSAYEPMIWKECKANSYLNWQWMRGIEDHVADFVIFHHQNWQ  65\n",
      "\n",
      "\n",
      "\tSequence 1 > 101 ------------------------------------ 101\n",
      "\t                                                         \n",
      "\tSequence 2 > 66  FVMIVIQMWAEMFEDDSICATARCWIYNRFLWSHRP 101\n",
      "\n",
      "\n",
      "-----------------------------------------------------------------------------------------------------\n",
      "Local Alignment: \n",
      "Alignment Score: 20.0\n",
      "Sequences Similarity: 100.0%\n",
      "\n",
      "\tSequence 1 > 97  HQ  98\n",
      "\t                 ||    \n",
      "\tSequence 2 > 61  HQ  62\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "p = Sequence_Alignment(mismatch=-10, gapOpen=40, gapExtend=2, diff_block=20)\n",
    "A = \"RFQLVEQLHWCGYNGEVTGCGQSQKMHRKPFNRRQVYNKAHMISLGVQPLQFQYRFTINYKSKWIVCNFDVPFWVENELPRADTRWAYSMIQKDVGHQAAV\"\n",
    "B = \"FTQQDEFVWLEWNDLRVKQEIEVDSAYEPMIWKECKANSYLNWQWMRGIEDHVADFVIFHHQNWQFVMIVIQMWAEMFEDDSICATARCWIYNRFLWSHRP\"\n",
    "\n",
    "print(\"GAP3 Alignment: \")\n",
    "p.gap3(A,B)\n",
    "print(\"-----------------------------------------------------------------------------------------------------\")\n",
    "print(\"Global Alignment: \")\n",
    "p.global_alignment(A,B)\n",
    "print(\"-----------------------------------------------------------------------------------------------------\")\n",
    "print(\"Local Alignment: \")\n",
    "p.local_alignment(A,B)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff5baa08",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
